File: backend/app/schemas/analytics.py
"""Pydantic schemas for the Phase 4 analytics API.

Two response families, matching the persisted-vs-live split described in
app/models/analytics.py and app/services/kpi_engine.py:

- Snapshot-backed reads (AnalyticsSummaryResponse, KPIsResponse,
  DataQualityReportRead) describe the most recent POST /analytics/run
  and never take filter query params - they're a frozen point-in-time
  read.
- Live-computed reads (AnomalyFindingRead, TrendPointRead,
  ProductRankingRead, DimensionPerformanceRead) are recomputed from the
  dataset's normalized rows on every call and do take filter params.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.analytics import AnalyticsJobStatus, KPIName
from app.models.dataset import FindingSeverity


class KPIResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    kpi_name: KPIName
    value: float
    unit: str
    computed_at: datetime


class DataQualityFindingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    severity: FindingSeverity
    category: str
    description: str
    recommendation: str
    created_at: datetime


class DataQualityReportRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    completeness_score: float
    validity_score: float
    consistency_score: float
    uniqueness_score: float
    overall_score: float
    created_at: datetime
    findings: list[DataQualityFindingRead]


class AnalyticsJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: AnalyticsJobStatus
    error_message: str | None
    started_at: datetime
    completed_at: datetime | None


class ChannelPerformanceRead(BaseModel):
    dimension: str
    units_sold: float | None = None
    revenue: float | None = None
    gross_profit: float | None = None
    units_returned: float | None = None
    return_rate: float | None = None


class AnalyticsSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    dataset_id: uuid.UUID
    job_id: uuid.UUID
    row_count: int
    column_count: int
    mapped_field_count: int
    summary: str | None
    created_at: datetime


class AnalyticsSummaryResponse(BaseModel):
    """GET /analytics/summary - the latest snapshot's headline figures:
    KPIs, the data-quality overall score, and channel performance (the
    one grouped metric from the spec's 25 with no endpoint of its own).
    """

    snapshot: AnalyticsSnapshotRead
    kpis: list[KPIResultRead]
    data_quality_overall_score: float | None
    channel_performance: list[ChannelPerformanceRead]


class KPIsResponse(BaseModel):
    snapshot: AnalyticsSnapshotRead
    kpis: list[KPIResultRead]


class RunAnalyticsResponse(BaseModel):
    job: AnalyticsJobRead
    snapshot: AnalyticsSnapshotRead


class AnomalyFindingRead(BaseModel):
    anomaly_type: str
    severity: str
    entity: str
    metric: str
    value: float
    z_score: float
    description: str


class AnomaliesResponse(BaseModel):
    dataset_id: uuid.UUID
    anomalies: list[AnomalyFindingRead]


class TrendPointRead(BaseModel):
    period: str
    units_sold: float | None = None
    revenue: float | None = None


class TrendsResponse(BaseModel):
    dataset_id: uuid.UUID
    granularity: str
    points: list[TrendPointRead]


class ProductRankingRead(BaseModel):
    product_id: str
    product_name: str | None = None
    units_sold: float | None = None
    revenue: float | None = None


class ProductsResponse(BaseModel):
    dataset_id: uuid.UUID
    top_products: list[ProductRankingRead]
    worst_products: list[ProductRankingRead]


class DimensionPerformanceRead(BaseModel):
    dimension: str
    units_sold: float | None = None
    revenue: float | None = None
    gross_profit: float | None = None
    units_returned: float | None = None
    return_rate: float | None = None


class CategoriesResponse(BaseModel):
    dataset_id: uuid.UUID
    categories: list[DimensionPerformanceRead]
    top_brands: list[DimensionPerformanceRead]


class SuppliersResponse(BaseModel):
    dataset_id: uuid.UUID
    suppliers: list[DimensionPerformanceRead]


class RegionsResponse(BaseModel):
    dataset_id: uuid.UUID
    regions: list[DimensionPerformanceRead]


class AnalyticsErrorDetail(BaseModel):
    """Structured error payload, mirroring DatasetErrorDetail so clients
    get a consistent shape across both dataset and analytics APIs."""

    code: str
    message: str

File: frontend/src/api/analytics.ts
import { apiFetch } from "./client";

export type KPIName =
  | "revenue"
  | "gross_profit"
  | "gross_margin"
  | "average_selling_price"
  | "average_order_value"
  | "inventory_value"
  | "inventory_turnover"
  | "sell_through_rate"
  | "return_rate"
  | "return_cost"
  | "units_sold"
  | "units_returned"
  | "stockouts"
  | "overstock_count"
  | "low_inventory_count";

export const KPI_LABELS: Record<KPIName, string> = {
  revenue: "Revenue",
  gross_profit: "Gross Profit",
  gross_margin: "Gross Margin",
  average_selling_price: "Avg. Selling Price",
  average_order_value: "Avg. Order Value",
  inventory_value: "Inventory Value",
  inventory_turnover: "Inventory Turnover",
  sell_through_rate: "Sell-Through Rate",
  return_rate: "Return Rate",
  return_cost: "Return Cost",
  units_sold: "Units Sold",
  units_returned: "Units Returned",
  stockouts: "Stockouts",
  overstock_count: "Overstock Count",
  low_inventory_count: "Low Inventory Count",
};

export interface KPIResult {
  kpi_name: KPIName;
  value: number;
  unit: string;
  computed_at: string;
}

export interface DataQualityFinding {
  severity: "info" | "warning" | "error";
  category: string;
  description: string;
  recommendation: string;
  created_at: string;
}

export interface DataQualityReport {
  completeness_score: number;
  validity_score: number;
  consistency_score: number;
  uniqueness_score: number;
  overall_score: number;
  created_at: string;
  findings: DataQualityFinding[];
}

export interface AnalyticsJob {
  id: string;
  status: "running" | "completed" | "failed";
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface AnalyticsSnapshot {
  id: string;
  dataset_id: string;
  job_id: string;
  row_count: number;
  column_count: number;
  mapped_field_count: number;
  summary: string | null;
  created_at: string;
}

export interface ChannelPerformance {
  dimension: string;
  units_sold: number | null;
  revenue: number | null;
  gross_profit: number | null;
  units_returned: number | null;
  return_rate: number | null;
}

export interface AnalyticsSummary {
  snapshot: AnalyticsSnapshot;
  kpis: KPIResult[];
  data_quality_overall_score: number | null;
  channel_performance: ChannelPerformance[];
}

export interface KPIsResponse {
  snapshot: AnalyticsSnapshot;
  kpis: KPIResult[];
}

export interface RunAnalyticsResponse {
  job: AnalyticsJob;
  snapshot: AnalyticsSnapshot;
}

export interface AnomalyFinding {
  anomaly_type: string;
  severity: string;
  entity: string;
  metric: string;
  value: number;
  z_score: number;
  description: string;
}

export interface AnomaliesResponse {
  dataset_id: string;
  anomalies: AnomalyFinding[];
}

export interface TrendPoint {
  period: string;
  units_sold: number | null;
  revenue: number | null;
}

export interface TrendsResponse {
  dataset_id: string;
  granularity: "daily" | "weekly" | "monthly";
  points: TrendPoint[];
}

export interface ProductRanking {
  product_id: string;
  product_name: string | null;
  units_sold: number | null;
  revenue: number | null;
}

export interface ProductsResponse {
  dataset_id: string;
  top_products: ProductRanking[];
  worst_products: ProductRanking[];
}

export interface DimensionPerformance {
  dimension: string;
  units_sold: number | null;
  revenue: number | null;
  gross_profit: number | null;
  units_returned: number | null;
  return_rate: number | null;
}

export interface CategoriesResponse {
  dataset_id: string;
  categories: DimensionPerformance[];
  top_brands: DimensionPerformance[];
}

export interface SuppliersResponse {
  dataset_id: string;
  suppliers: DimensionPerformance[];
}

export interface RegionsResponse {
  dataset_id: string;
  regions: DimensionPerformance[];
}

export interface AnalyticsFilters {
  dateFrom?: string;
  dateTo?: string;
  category?: string;
  supplier?: string;
  region?: string;
  channel?: string;
}

/**
 * Error raised by every function in this module. Analytics endpoints
 * return a structured `{code, message}` error body (see
 * app/schemas/analytics.py AnalyticsErrorDetail).
 */
export class AnalyticsApiError extends Error {
  status: number;
  code: string;

  constructor(status: number, code: string, message: string) {
    super(message);
    this.name = "AnalyticsApiError";
    this.status = status;
    this.code = code;
  }
}

function analyticsErrorFromBody(status: number, body: unknown): AnalyticsApiError {
  const detail = (body as { detail?: unknown } | null)?.detail;
  if (detail && typeof detail === "object") {
    const d = detail as { code?: string; message?: string };
    if (typeof d.message === "string") {
      return new AnalyticsApiError(status, d.code ?? "UNKNOWN_ERROR", d.message);
    }
  }
  if (typeof detail === "string") {
    return new AnalyticsApiError(status, "UNKNOWN_ERROR", detail);
  }
  return new AnalyticsApiError(status, "UNKNOWN_ERROR", `Request failed with status ${status}`);
}

function buildQuery(
  datasetId: string,
  filters: AnalyticsFilters = {},
  extra: Record<string, string | number | undefined> = {},
): string {
  const params = new URLSearchParams();
  params.set("dataset_id", datasetId);
  if (filters.dateFrom) params.set("date_from", filters.dateFrom);
  if (filters.dateTo) params.set("date_to", filters.dateTo);
  if (filters.category) params.set("category", filters.category);
  if (filters.supplier) params.set("supplier", filters.supplier);
  if (filters.region) params.set("region", filters.region);
  if (filters.channel) params.set("channel", filters.channel);
  for (const [key, value] of Object.entries(extra)) {
    if (value !== undefined) params.set(key, String(value));
  }
  return params.toString();
}

async function analyticsFetchJson<T>(path: string): Promise<T> {
  const response = await apiFetch(path);
  if (!response.ok) {
    let body: unknown = null;
    try {
      body = await response.clone().json();
    } catch {
      // response wasn't JSON; analyticsErrorFromBody falls back to a generic message.
    }
    throw analyticsErrorFromBody(response.status, body);
  }
  return (await response.json()) as T;
}

export async function runAnalytics(datasetId: string): Promise<RunAnalyticsResponse> {
  const response = await apiFetch(`/analytics/run?dataset_id=${encodeURIComponent(datasetId)}`, {
    method: "POST",
  });
  if (!response.ok) {
    let body: unknown = null;
    try {
      body = await response.clone().json();
    } catch {
      // response wasn't JSON; analyticsErrorFromBody falls back to a generic message.
    }
    throw analyticsErrorFromBody(response.status, body);
  }
  return (await response.json()) as RunAnalyticsResponse;
}

export async function getAnalyticsSummary(datasetId: string): Promise<AnalyticsSummary> {
  return analyticsFetchJson<AnalyticsSummary>(`/analytics/summary?${buildQuery(datasetId)}`);
}

export async function getKPIs(datasetId: string): Promise<KPIsResponse> {
  return analyticsFetchJson<KPIsResponse>(`/analytics/kpis?${buildQuery(datasetId)}`);
}

export async function getDataQuality(datasetId: string): Promise<DataQualityReport> {
  return analyticsFetchJson<DataQualityReport>(`/analytics/data-quality?${buildQuery(datasetId)}`);
}

export async function getAnomalies(
  datasetId: string,
  filters: AnalyticsFilters = {},
): Promise<AnomaliesResponse> {
  return analyticsFetchJson<AnomaliesResponse>(
    `/analytics/anomalies?${buildQuery(datasetId, filters)}`,
  );
}

export async function getTrends(
  datasetId: string,
  granularity: "daily" | "weekly" | "monthly" = "monthly",
  filters: AnalyticsFilters = {},
): Promise<TrendsResponse> {
  const query = buildQuery(datasetId, filters, { granularity });
  return analyticsFetchJson<TrendsResponse>(`/analytics/trends?${query}`);
}

export async function getProducts(
  datasetId: string,
  filters: AnalyticsFilters = {},
  limit?: number,
): Promise<ProductsResponse> {
  const query = buildQuery(datasetId, filters, { limit });
  return analyticsFetchJson<ProductsResponse>(`/analytics/products?${query}`);
}

export async function getCategories(
  datasetId: string,
  filters: AnalyticsFilters = {},
  limit?: number,
): Promise<CategoriesResponse> {
  const query = buildQuery(datasetId, filters, { limit });
  return analyticsFetchJson<CategoriesResponse>(`/analytics/categories?${query}`);
}

export async function getSuppliers(
  datasetId: string,
  filters: AnalyticsFilters = {},
  limit?: number,
): Promise<SuppliersResponse> {
  const query = buildQuery(datasetId, filters, { limit });
  return analyticsFetchJson<SuppliersResponse>(`/analytics/suppliers?${query}`);
}

export async function getRegions(
  datasetId: string,
  filters: AnalyticsFilters = {},
  limit?: number,
): Promise<RegionsResponse> {
  const query = buildQuery(datasetId, filters, { limit });
  return analyticsFetchJson<RegionsResponse>(`/analytics/regions?${query}`);
}
